File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
//given an array, find the first duplicate and return its value. If no duplicate is found, return -1

function findFirstDuplicate(arr) {
  // use a set to track the seen values
  let hash = new Set();

  for (let i = 0; i < arr.length; i++) {
    //if we have seen the current value, return it
    if (hash.has(arr[i])) {
      return arr[i];
    }
    //otherwise, add it to the set
    hash.add(arr[i]);
  }

  //if we never return a value in the loop, return -1
  return -1;
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
// And a written explanation of your solution

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
def find_first_duplicate(arr)
  # type your code in here
  hash_map = Hash.new

  arr.each do |el|
    if hash_map.has_key?(el) 
      return el
    else
      hash_map[el] = 1
    end
  end

  return -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 2"
  puts "=>", find_first_duplicate([2, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  puts
  
  puts "Expecting: 4"
  puts "=>", find_first_duplicate([2, 4, 5, 4, 1, 2])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution
